Shared/ViewModels/ItemDetailsViewModel/ItemDetailsViewModel.swift

```diff
@@ -0,0 +1,185 @@
+//
+
+import Combine
+import Foundation
+import JellyfinAPI
+
+class ItemDetailsViewModel<ItemType: Equatable>: ViewModel, Stateful, Eventful {
+
+    // MARK: - Events
+
+    enum Event: Equatable {
+        case updated
+        case error(JellyfinAPIError)
+    }
+
+    // MARK: - Actions
+
+    enum Action: Equatable {
+        case add([ItemType])
+        case remove([ItemType])
+        case update(BaseItemDto)
+    }
+
+    // MARK: - State
+
+    enum State: Hashable {
+        case initial
+        case error(JellyfinAPIError)
+        case updating
+        case refreshing
+    }
+
+    @Published
+    var item: BaseItemDto
+
+    @Published
+    var state: State = .initial
+
+    private var updateTask: AnyCancellable?
+    private let eventSubject = PassthroughSubject<Event, Never>()
+
+    var events: AnyPublisher<Event, Never> {
+        eventSubject.receive(on: RunLoop.main).eraseToAnyPublisher()
+    }
+
+    // MARK: - Init
+
+    init(item: BaseItemDto) {
+        self.item = item
+        super.init()
+    }
+
+    // MARK: - Respond to Actions
+
+    func respond(to action: Action) -> State {
+        switch action {
+        case let .add(items):
+            return updateStateAndPerform(.updated) {
+                try await self.addItems(items)
+            }
+        case let .remove(items):
+            return updateStateAndPerform(.updated) {
+                try await self.removeItems(items)
+            }
+        case let .update(item):
+            return updateStateAndPerform(.updated) {
+                _ = self.updateItem(item)
+            }
+        }
+    }
+
+    // MARK: - Update State and Perform Operation
+
+    private func updateStateAndPerform(_ event: Event, operation: @escaping () async throws -> Void) -> State {
+        cancelCurrentTask()
+
+        updateTask = Task { [weak self] in
+            guard let self = self else { return }
+            await self.setState(.updating)
+
+            do {
+                try await operation()
+                await self.completeWithSuccess(event)
+            } catch {
+                await self.handleTaskError(error)
+            }
+        }.asAnyCancellable()
+
+        return .refreshing
+    }
+
+    // MARK: - Cancel Current Task
+
+    private func cancelCurrentTask() {
+        updateTask?.cancel()
+    }
+
+    // MARK: - Complete with Success
+
+    private func completeWithSuccess(_ event: Event) async {
+        await setState(.initial)
+        eventSubject.send(event)
+    }
+
+    // MARK: - Handle Task Error
+
+    private func handleTaskError(_ error: Error) async {
+        guard !Task.isCancelled else { return }
+        let apiError = JellyfinAPIError(error.localizedDescription)
+        await setState(.error(apiError))
+        eventSubject.send(.error(apiError))
+    }
+
+    // MARK: - Update Item on Server
+
+    func updateItem(_ newItem: BaseItemDto, refresh: Bool = false) -> State {
```

Comment:
What would be the reason we would want to refresh locally vs on the server? I see that we do so differently based on the genres, people, etc.

Reply:
Since the Genre & Tag are just [String] updating `item = newItem` just works. However, the person/studio is being input as a name. Jellyfin turns these into full BaseItemPerson/NameGuidPair. I've tried updating it as `item = newItem` but the list view isn't able to get the image for the person and the Studio doesn't actually link to a Studio object since it's just `guid=nil` and `name=name`. 

The only reason I'm not updating the Genre and Tag is it felt unnecessary. 

The other route for this is to make the addPeople/addStudio view look up the item instead of just inputting from a name. I'd be game to make those lookups on the server but I might need some guidance on what we'd what that view to look like.